### ISSUE TITLE
Distance update delay overfiltered for slower intervals

### ISSUE THREAD

**Issue by kentestforjob** [ESPresense/ESPresense#770]
### Describe the bug

Config of Espresense: Default
ESPresense Version: v3.1.1
ESP: ESP-WROOM-32 ([link](https://amzn.to/3xCAw1F))

**Situation**:
At initial, my iphone or watch close the ESP32 around 0.2m. Then walk away it to 1m just in 1 second.
But the response show that the distance still keep in 0.2m, 0.2xm,  0.2xm, 0.3m.....1m (it takes around 8-20s to adjust it and finally it get the correct distance)

**Observation**: 
1) The rssi(-32) > rssi@1m(-65) (e.g ) when my device near the ESP around 0.2m

2) The rssi(-65) ~= rssi@1m(-65) (e.g ) when my device move away from the ESP around 1m (change it significantly )

3) The raw(0.9~1.9) when my device near the ESP around 1m (change it significantly but unstable)

**Conclusion/ Question/ Problem**:
RSSI or raw can reflect the latest status of my device (although I don't know what is "raw", i think it likely similar to distance but it is unstable). And It is strange that why "distance" can not reflect the current status and it needs time to change it slowly (0.2 -> 0.2x, 0.3 -> 0.3x -> ....-> 1). Suppose the formula of distance is base on rssi.

Anyone have such problem or anything I need to config it?

### Version

v3.1.1

### Active scan enabled

no

### Include filter

no

### Exclude filter

no

### Query filter

no

### Device (be specific)

iphone12, iosWatch

### Logs

```text
10:19:52
{"id":"irk:____","disc":"071897c6e5","idType":200,"rssi@1m":-65,"rssi":-34,"raw":0.13,"distance":0.15,"speed":-0.00,"mac":"430e33182fc7","interval":492}
{"id":"irk:____","disc":"071897c6e5","idType":200,"rssi@1m":-65,"rssi":-41,"raw":0.21,"distance":0.15,"speed":0.00,"mac":"430e33182fc7","interval":491}
{"id":"irk:____","disc":"071897c6e5","idType":200,"rssi@1m":-65,"rssi":-52,"raw":0.43,"distance":0.15,"speed":0.01,"mac":"430e33182fc7","interval":491}
10:19:54
{"id":"irk:____","disc":"071897c6e5","idType":200,"rssi@1m":-65,"rssi":-72,"raw":1.58,"distance":0.16,"speed":0.06,"mac":"430e33182fc7","interval":491}
{"id":"irk:____","disc":"071897c6e5","idType":200,"rssi@1m":-65,"rssi":-72,"raw":1.58,"distance":0.20,"speed":0.09,"mac":"430e33182fc7","interval":491}
{"id":"irk:____","disc":"071897c6e5","idType":200,"rssi@1m":-65,"rssi":-69,"raw":1.30,"distance":0.26,"speed":0.06,"mac":"430e33182fc7","interval":489}
{"id":"irk:____","disc":"071897c6e5","idType":200,"rssi@1m":-65,"rssi":-70,"raw":1.39,"distance":0.28,"speed":0.06,"mac":"430e33182fc7","interval":492}
{"id":"irk:____","disc":"071897c6e5","idType":200,"rssi@1m":-65,"rssi":-66,"raw":1.07,"distance":0.36,"speed":0.06,"mac":"430e33182fc7","interval":490}
{"id":"irk:____","disc":"071897c6e5","idType":200,"rssi@1m":-65,"rssi":-64,"raw":0.94,"distance":0.37,"speed":0.04,"mac":"430e33182fc7","interval":491}
{"id":"irk:____","disc":"071897c6e5","idType":200,"rssi@1m":-65,"rssi":-64,"raw":0.94,"distance":0.37,"speed":0.04,"mac":"430e33182fc7","interval":491}
{"id":"irk:____","disc":"071897c6e5","idType":200,"rssi@1m":-65,"rssi":-63,"raw":0.88,"distance":0.40,"speed":0.03,"mac":"430e33182fc7","interval":493}
{"id":"irk:____","disc":"071897c6e5","idType":200,"rssi@1m":-65,"rssi":-66,"raw":1.07,"distance":0.52,"speed":0.04,"mac":"430e33182fc7","interval":492}
{"id":"irk:____","disc":"071897c6e5","idType":200,"rssi@1m":-65,"rssi":-68,"raw":1.22,"distance":0.53,"speed":0.04,"mac":"430e33182fc7","interval":492}
{"id":"irk:____","disc":"071897c6e5","idType":200,"rssi@1m":-65,"rssi":-78,"raw":2.35,"distance":0.61,"speed":0.07,"mac":"430e33182fc7","interval":494}
{"id":"irk:____","disc":"071897c6e5","idType":200,"rssi@1m":-65,"rssi":-74,"raw":1.81,"distance":0.70,"speed":0.09,"mac":"430e33182fc7","interval":497}
{"id":"irk:____","disc":"071897c6e5","idType":200,"rssi@1m":-65,"rssi":-74,"raw":1.81,"distance":0.81,"speed":0.06,"mac":"430e33182fc7","interval":495}
{"id":"irk:____","disc":"071897c6e5","idType":200,"rssi@1m":-65,"rssi":-69,"raw":1.30,"distance":0.85,"speed":0.03,"mac":"430e33182fc7","interval":496}
{"id":"irk:____","disc":"071897c6e5","idType":200,"rssi@1m":-65,"rssi":-69,"raw":1.30,"distance":0.90,"speed":0.02,"mac":"430e33182fc7","interval":498}
10:20:13
{"id":"irk:____","disc":"071897c6e5","idType":200,"rssi@1m":-65,"rssi":-75,"raw":1.93,"distance":1.00,"speed":0.05,"mac":"430e33182fc7","interval":496}
```

**Comment by DTTerastar** [comment]
Rssi is a noisy signal so post processing is done to try and filter out noise.  I'd love someone to work on that piece of code and see if we can improve responsiveness without increasing noise!

**Comment by kentestforjob** [comment]
Thanks your reply @DTTerastar 

In my log show that "rssi@1m":-65,"rssi":-66,"raw":1.07,"distance":0.52
rssi@1m iis -65, and the rssi is -66 but the distance is 0.52. Is it incorrect? Why `raw` is more likely a calculated distance.

**Comment by liamrpowell** [comment]
I noticed this too, especially when walking into a room. It takes a good 15 seconds to register that I am in the room and run the light automation I have setup.

**Comment by liamrpowell** [comment]
> Yeah, that's over-filtering... I tuned it for the apple watch interval of 500ms, and it's slow for everything else...

I trie 500ms in filter and still takes 15 seconds to respond

**Comment by theabian** [comment]
> I trie 500ms in filter and still takes 15 seconds to respond

Did you find an solution for that problem?

**Comment by liamrpowell** [comment]
Nope, nothing yet.Sent from my iPhoneOn Mar 30, 2023, at 7:18 PM, theabian ***@***.***> wrote:﻿

I trie 500ms in filter and still takes 15 seconds to respond

Did you find an solution for that problem?

**Comment by theabian** [comment]
When using the option "Maximum distance to report" it takes the esp ages to "wake" up and report the device nearby

When disabling the option, there is already the next problem, the distance value is taking ages to update.

Wish I could use the "raw" value instand of the "distance" value, in my case it updates pretty fast and reliable

**Comment by kentestforjob** [comment]
@theabian 
Yes, I am using "raw" instead of "distance" but the value may be drop or increase suddenly and the gap of value may be big that is another problem.

@liamrpowell 
I tried to setup the "Included devices with age less than (in ms) to 500" in Counting and "Skip reporting if message age is less that this in ms 500" in Filtering, then the response time will be faster ~ less than 1 second. It may carry out another problem if there are a lot of Bluetooth devices in surrounding, the ESP32 and the program sometime maybe will crash. I think the computing power is not enough to handle such many devices.
( **However, the delay of distance still same as before**. )

**Comment by theabian** [comment]
> @theabian Yes, I am using "raw" instead of "distance" but the value may be drop or increase suddenly and the gap of value may be big that is another problem.
> 
> @liamrpowell I tried to setup the "Included devices with age less than (in ms) to 500" in Counting and "Skip reporting if message age is less that this in ms 500" in Filtering, then the response time will be faster ~ less than 1 second. It may carry out another problem if there are a lot of Bluetooth devices in surrounding, the ESP32 and the program sometime maybe will crash. I think the computing power is not enough to handle such many devices. ( **However, the delay of distance still same as before**. )

I´m now at this point too.
In my case the ESP32/program never crashed.

Hope the "delay of distance" gets fixed or someone can give us an hint what we need to configure.

**Comment by johnnykudahl** [comment]
I have not super deep dived into the code, but see this in blefingerprint

```
bool BleFingerprint::seen(BLEAdvertisedDevice *advertisedDevice) {
    lastSeenMillis = millis();
    reported = false;

    seenCount++;

    fingerprint(advertisedDevice);

    if (ignore) return false;

    oldest = recent;
    recent = newest;
    newest = advertisedDevice->getRSSI();

    auto the_min = min(min(oldest, recent), newest);
    auto the_max = max(max(oldest, recent), newest);
    auto the_median = the_max ^ the_min ^ oldest ^ recent ^ newest;
    auto range = the_max - the_min;

    rssi = the_median;

    const auto ratio = float(get1mRssi() - rssi) / (10.0f * BleFingerprintCollection::absorption);
    raw = pow(10, ratio);
    if (filter()) hasValue = true; 
```

This formular make sense (that is how you calualate distance usign rssi values)
```
 const auto ratio = float(get1mRssi() - rssi) / (10.0f * BleFingerprintCollection::absorption);
    raw = pow(10, ratio);
```

But getting the media by bitwise XOR'ing the RSSI values I do not understand ...
`auto the_median = the_max ^ the_min ^ oldest ^ recent ^ newest;`

**Comment by DTTerastar** [comment]
I agree that is weird looking.  But xor basically flips bits if set so it's just a fancy way finding the middle value because we xor 3 values then xor the biggest and littlest which leaves the middle.

**Comment by johnnykudahl** [comment]
^ is bitwise xor right?

So e.g.
the_max = 50 (00110010)
the_min = 40 (00101000)
oldest =  42 (00101010)
recent =  44 (00101100)
newest =  48 (00110000)
average = 44.8

00110010
00101000
00101010
00101100
00110000
xor’ed
00111110 => 62 decimal

**Comment by DTTerastar** [comment]
the_max needs to actually be 48, and the min needs to be 42 then it should work

it basically "erases" those values from the combined value, get it?

**Comment by johnnykudahl** [comment]
Made a mistake in my xor earlier common, sorry for that and when I corrected that the penny dropped… It is the median and not mean value ... xor will give the median, true.

**Comment by maranhaoBruno** [comment]
Is anybody involved in the development of ESPresense following this issue? It’s kind of insane that when I walk from my bedroom where my phone is reading 15 meters away from my kitchen beacon to less than one meter away from my kitchen beacon it takes over 30 seconds for the distance to fall to within 2 meters despite it taking me 5 seconds to make the walk. It makes automating based on walking into the kitchen impossible. It would be nice to control how fast or slow the distance variable responds to RSSI changes.

**Comment by DTTerastar** [comment]
The beta has a fix for this, please test.

**Comment by maranhaoBruno** [comment]
Any idea when that fix will be pushed to non-beta?

**Comment by DTTerastar** [comment]
Once I know it's stable and won't brick people's nodes...

**Comment by angristan** [comment]
> The beta has a fix for this, please test.

I am testing [the fix](https://github.com/ESPresense/ESPresense/pull/1108) by using 3.3.2b3, I can report that `distance` is indeed quicker to adjust following a sharp increase/drop in RSSI!
